### PR TITLE
Implement circuit breaker with Spring Cloud Netflix Hystrix

### PR DESCRIPTION
**Pull request by arey**
In the `api-gateway`module, I've added a circuit breaker to the `VisitsServiceClient::getVisitsForPets` method. When the `pets-service` is down (or not started), the owner's detail page is now displayed without the visits.
I'm aware about all other methods where we could apply the circuit breaker pattern.

Metrics are available at the URL: http://localhost:8080/actuator/hystrix.stream

We could add a Hystrix Dashboard in another Maven module. I've tried to add it to the `api-gaeway' module but I had some conflicts with the jQuery webjar.

### CODE REVIEW

**Comment by mszarlinski** [comment]
Thanks for PR, please answer my comments. I think Hystrix Turbine/Dashboard should be deployed independently.

File: spring-petclinic-api-gateway/src/main/java/org/springframework/samples/petclinic/api/application/VisitsServiceClient.java [hunk 1]
```diff
@@ -51,4 +55,8 @@
     private String joinIds(List<Integer> petIds) {
         return petIds.stream().map(Object::toString).collect(joining(","));
     }
+
+    public Map<Integer, List<VisitDetails>> emptyVisitsForPets(List<Integer> petIds) {
+        return new HashMap<>();
```
[the remark this answers is not in the document]
Comment (arey) [reply spring-petclinic-api-gateway/src/main/java/org/springframework/samples/petclinic/api/application/VisitsServiceClient.java]: You're right

File: spring-petclinic-api-gateway/src/main/java/org/springframework/samples/petclinic/api/application/VisitsServiceClient.java [hunk 2]
```diff
@@ -51,4 +55,8 @@
     private String joinIds(List<Integer> petIds) {
         return petIds.stream().map(Object::toString).collect(joining(","));
     }
+
+    public Map<Integer, List<VisitDetails>> emptyVisitsForPets(List<Integer> petIds) {
```
[the remark this answers is not in the document]
Comment (mszarlinski) [reply spring-petclinic-api-gateway/src/main/java/org/springframework/samples/petclinic/api/application/VisitsServiceClient.java]: Crazy reflection :)

File: spring-petclinic-api-gateway/src/main/java/org/springframework/samples/petclinic/api/application/VisitsServiceClient.java [hunk 3]
```diff
@@ -38,6 +41,7 @@
 
     private final RestTemplate loadBalancedRestTemplate;
 
+    @HystrixCommand(fallbackMethod = "emptyVisitsForPets")
```
[the remark this answers is not in the document]
Comment (arey) [reply spring-petclinic-api-gateway/src/main/java/org/springframework/samples/petclinic/api/application/VisitsServiceClient.java]: Sure. MockRestServiceServer will do the job.

**Comment by arey** [comment]
> Thanks for PR, please answer my comments. I think Hystrix Turbine/Dashboard should be deployed independently.

I think so. Maybe in another PR?

File: spring-petclinic-api-gateway/src/test/java/org/springframework/samples/petclinic/api/application/VisitsServiceClientTest.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    @Test
+    public void getVisitsForPets_withServerError() {
+
+        mockServer.expect(requestTo("http://visits-service/pets/visits?petId=1"))
+            .andRespond(withServerError());
+
+        Map<Integer, List<VisitDetails>> visits = visitsServiceClient.getVisitsForPets(Collections.singletonList(1));
+
+        assertNotNull(visits);
+        assertEquals(0, visits.size());
```
Comment (mszarlinski) [inline spring-petclinic-api-gateway/src/test/java/org/springframework/samples/petclinic/api/application/VisitsServiceClientTest.java]: I think first assertion is redundant

File: spring-petclinic-api-gateway/src/test/java/org/springframework/samples/petclinic/api/application/VisitsServiceClientTest.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        mockServer.expect(requestTo("http://visits-service/pets/visits?petId=1"))
+            .andRespond(withSuccess("{\"items\":[{\"id\":5,\"date\":\"2018-11-15\",\"description\":\"test visit\",\"petId\":1}]}", MediaType.APPLICATION_JSON));
+
+        Map<Integer, List<VisitDetails>> visits = visitsServiceClient.getVisitsForPets(Collections.singletonList(1));
+
+        assertNotNull(visits);
+        assertEquals(1, visits.size());
+        assertNotNull(visits.get(1));
+        assertEquals(1, visits.get(1).size());
+        assertEquals("test visit", visits.get(1).get(0).getDescription());
```
Comment (mszarlinski) [inline spring-petclinic-api-gateway/src/test/java/org/springframework/samples/petclinic/api/application/VisitsServiceClientTest.java]: I would rather extract it to custom private asserion method and extract constant/variable: assertDescriptionEquals(visits.get(PET_ID) ‚’’test visit’’)

Comment (arey) [reply spring-petclinic-api-gateway/src/test/java/org/springframework/samples/petclinic/api/application/VisitsServiceClientTest.java]: I've done the change. Is-it what you expect?

File: spring-petclinic-api-gateway/src/test/java/org/springframework/samples/petclinic/api/application/VisitsServiceClientTest.java (hunk 4 above)
Comment (mszarlinski) [reply spring-petclinic-api-gateway/src/test/java/org/springframework/samples/petclinic/api/application/VisitsServiceClientTest.java]: If map is `null`, then `assertEquals(0, visits.size());` signals error and fails a build. Less assertions - less noise in tests.

File: spring-petclinic-api-gateway/src/test/java/org/springframework/samples/petclinic/api/application/VisitsServiceClientTest.java (hunk 5 above)
Comment (mszarlinski) [reply spring-petclinic-api-gateway/src/test/java/org/springframework/samples/petclinic/api/application/VisitsServiceClientTest.java]: I suggest to hide **all** assertions in new custom method. Then we have one meaningful assertion in test, so it becomes easier to read and understand.

File: spring-petclinic-api-gateway/src/test/java/org/springframework/samples/petclinic/api/application/VisitsServiceClientTest.java (hunk 4 above)
Comment (arey) [reply spring-petclinic-api-gateway/src/test/java/org/springframework/samples/petclinic/api/application/VisitsServiceClientTest.java]: That's a point of view. I prefer to fail the test as soon as possible. But if done the change.

File: spring-petclinic-api-gateway/src/test/java/org/springframework/samples/petclinic/api/application/VisitsServiceClientTest.java (hunk 5 above)
Comment (arey) [reply spring-petclinic-api-gateway/src/test/java/org/springframework/samples/petclinic/api/application/VisitsServiceClientTest.java]: I've refactored the test. It should be ok.

**Pull request merged**